lib/src/http/auto_retry.dart

```diff
@@ -0,0 +1,105 @@
+library w_transport.src.http.auto_retry;
+
+import 'package:w_transport/src/http/base_request.dart';
+import 'package:w_transport/src/http/finalized_request.dart';
+import 'package:w_transport/src/http/request_exception.dart';
+import 'package:w_transport/src/http/requests.dart';
+import 'package:w_transport/src/http/response.dart';
+
+typedef /* Future<bool> | bool */ RetryTest(
+    FinalizedRequest request, BaseResponse response);
+
+/// Deciding whether or not to retry a failed request is determined by the
+/// settings defined in fields in this class.
+///
+/// A request will be retried if all of the following are true:
+///
+/// - [enabled] is true
+/// - [forHttpMethods] includes the HTTP method that the failed request used
+/// - [forStatusCodes] includes the response's status code
+/// - [test] (if not null) returns true
```

Comment:
we should chat through the logic of this set of retry configurations.  it may impose too much configuration for simple use cases.  if all conditions must be met, does that mean you have to ensure that at least 3 of these things are properly configured?